File: src-tauri/src/fs.rs
use serde::Serialize;
use std::fs;
use std::path::Path;

#[derive(Debug, Clone, Serialize)]
pub struct TreeNode {
    pub name: String,
    pub path: String,
    #[serde(rename = "type")]
    pub node_type: String, // "file" or "dir"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub children: Option<Vec<TreeNode>>,
}

fn build_tree(path: &Path, base: &Path) -> Vec<TreeNode> {
    let mut nodes = Vec::new();

    let entries = match fs::read_dir(path) {
        Ok(entries) => entries,
        Err(_) => return nodes,
    };

    let mut items: Vec<_> = entries.flatten().collect();
    items.sort_by_key(|e| e.file_name());

    for entry in items {
        let entry_path = entry.path();
        let name = entry.file_name().to_string_lossy().to_string();

        // Skip hidden files/dirs
        if name.starts_with('.') {
            continue;
        }

        let relative = entry_path
            .strip_prefix(base)
            .unwrap_or(&entry_path)
            .to_string_lossy()
            .to_string();

        if entry_path.is_dir() {
            let children = build_tree(&entry_path, base);
            nodes.push(TreeNode {
                name,
                path: relative,
                node_type: "dir".to_string(),
                children: Some(children),
            });
        } else {
            nodes.push(TreeNode {
                name,
                path: relative,
                node_type: "file".to_string(),
                children: None,
            });
        }
    }

    nodes
}

#[tauri::command]
pub fn read_file(path: String) -> Result<String, String> {
    fs::read_to_string(&path).map_err(|e| format!("Failed to read file: {}", e))
}

#[tauri::command]
pub fn read_dir_tree(path: String) -> Result<Vec<TreeNode>, String> {
    let p = Path::new(&path);
    if !p.is_dir() {
        return Err(format!("Not a directory: {}", path));
    }
    Ok(build_tree(p, p))
}

#[tauri::command]
pub fn path_exists(path: String) -> bool {
    Path::new(&path).exists()
}

const FAVICON_CANDIDATES: &[&str] = &[
    "favicon.svg",
    "favicon.ico",
    "favicon.png",
    "public/favicon.svg",
    "public/favicon.ico",
    "public/favicon.png",
    "app/favicon.ico",
    "app/favicon.png",
    "app/icon.svg",
    "app/icon.png",
    "app/icon.ico",
    "src/favicon.ico",
    "src/favicon.svg",
    "src/app/favicon.ico",
    "src/app/icon.svg",
    "src/app/icon.png",
    "assets/icon.svg",
    "assets/icon.png",
    "assets/logo.svg",
    "assets/logo.png",
];

fn mime_for_ext(ext: &str) -> &'static str {
    match ext {
        "svg" => "image/svg+xml",
        "png" => "image/png",
        "ico" => "image/x-icon",
        "jpg" | "jpeg" => "image/jpeg",
        _ => "application/octet-stream",
    }
}

#[tauri::command]
pub fn find_project_favicon(project_path: String) -> Option<String> {
    use base64::Engine;
    let base = Path::new(&project_path);
    for candidate in FAVICON_CANDIDATES {
        let p = base.join(candidate);
        if p.is_file() {
            if let Ok(bytes) = fs::read(&p) {
                let ext = p
                    .extension()
                    .and_then(|e| e.to_str())
                    .unwrap_or("png");
                let mime = mime_for_ext(ext);
                let b64 = base64::engine::general_purpose::STANDARD.encode(&bytes);
                return Some(format!("data:{};base64,{}", mime, b64));
            }
        }
    }
    None
}

#[tauri::command]
pub fn list_dir(path: String) -> Result<Vec<String>, String> {
    let p = Path::new(&path);
    if !p.is_dir() {
        return Err(format!("Not a directory: {}", path));
    }
    let mut names = Vec::new();
    let entries = fs::read_dir(p).map_err(|e| format!("Failed to read dir: {}", e))?;
    for entry in entries.flatten() {
        names.push(entry.file_name().to_string_lossy().to_string());
    }
    names.sort();
    Ok(names)
}

File: src/main/ipc/fs.ts
import fs from 'node:fs'
import path from 'node:path'
import { ipcMain } from 'electron'

interface TreeNode {
  name: string
  path: string
  type: 'file' | 'dir'
  children?: TreeNode[]
}

function buildTree(dirPath: string, basePath: string): TreeNode[] {
  const nodes: TreeNode[] = []

  let entries: fs.Dirent[]
  try {
    entries = fs.readdirSync(dirPath, { withFileTypes: true })
  } catch {
    return nodes
  }

  entries.sort((a, b) => a.name.localeCompare(b.name))

  for (const entry of entries) {
    // Skip hidden files/dirs
    if (entry.name.startsWith('.')) continue

    const fullPath = path.join(dirPath, entry.name)
    const relativePath = path.relative(basePath, fullPath)

    if (entry.isDirectory()) {
      nodes.push({
        name: entry.name,
        path: relativePath,
        type: 'dir',
        children: buildTree(fullPath, basePath),
      })
    } else {
      nodes.push({
        name: entry.name,
        path: relativePath,
        type: 'file',
      })
    }
  }

  return nodes
}

const FAVICON_CANDIDATES = [
  'favicon.svg',
  'favicon.ico',
  'favicon.png',
  'public/favicon.svg',
  'public/favicon.ico',
  'public/favicon.png',
  'app/favicon.ico',
  'app/favicon.png',
  'app/icon.svg',
  'app/icon.png',
  'app/icon.ico',
  'src/favicon.ico',
  'src/favicon.svg',
  'src/app/favicon.ico',
  'src/app/icon.svg',
  'src/app/icon.png',
  'assets/icon.svg',
  'assets/icon.png',
  'assets/logo.svg',
  'assets/logo.png',
]

function mimeForExt(ext: string): string {
  switch (ext) {
    case 'svg': return 'image/svg+xml'
    case 'png': return 'image/png'
    case 'ico': return 'image/x-icon'
    case 'jpg':
    case 'jpeg': return 'image/jpeg'
    default: return 'application/octet-stream'
  }
}

ipcMain.handle('read_file', (_event, args: { path: string }): string => {
  return fs.readFileSync(args.path, 'utf-8')
})

ipcMain.handle('read_dir_tree', (_event, args: { path: string }): TreeNode[] => {
  const p = args.path
  if (!fs.existsSync(p) || !fs.statSync(p).isDirectory()) {
    throw new Error(`Not a directory: ${p}`)
  }
  return buildTree(p, p)
})

ipcMain.handle('path_exists', (_event, args: { path: string }): boolean => {
  return fs.existsSync(args.path)
})

ipcMain.handle('list_dir', (_event, args: { path: string }): string[] => {
  const p = args.path
  if (!fs.existsSync(p) || !fs.statSync(p).isDirectory()) {
    throw new Error(`Not a directory: ${p}`)
  }
  const entries = fs.readdirSync(p)
  entries.sort()
  return entries
})

ipcMain.handle(
  'find_project_favicon',
  (_event, args: { projectPath: string }): string | null => {
    for (const candidate of FAVICON_CANDIDATES) {
      const fullPath = path.join(args.projectPath, candidate)
      if (fs.existsSync(fullPath) && fs.statSync(fullPath).isFile()) {
        try {
          const bytes = fs.readFileSync(fullPath)
          const ext = path.extname(fullPath).slice(1) || 'png'
          const mime = mimeForExt(ext)
          const b64 = bytes.toString('base64')
          return `data:${mime};base64,${b64}`
        } catch {
          continue
        }
      }
    }
    return null
  },
)
